Pynumero interior point

michaelbynum opened the pull request:
@Robbybp There is still work to do, but I'd like your input on the changes I have been making. I tried to simplify a lot of code (both in the `InteriorPointSolver` and the `InteriorPointInterface`). 
1. I added the `LinearSolverResults` object as we discussed. 
2. I removed all of the compression matrix stuff to improve performance. 
3. I was having trouble with the the second half of the regularization test. I was getting errors because the regularization coefficient was getting to large, so I cut the second half of that test (the first half was perfect), and added a second test.
4. I modified the logging slightly to log every iteration of the factorization rather than just iterations with regularization (still with debug).

emma58 committed:
Merging name buffer changes, incomplete changes to mappings

emma58 committed:
Making disaggregated var map and constraint maps consistent with each other, changing tests to match

emma58 committed:
Adding mapping from variables and disjunctions to the disaggregations constraints

emma58 committed:
Adding a test of a pathological case where a variable is global because of the objective, updating chull tests, fixing bug with indexed block on a disjunction in chull

emma58 committed:
Fixing a typo

emma58 committed:
Fixing chull to not move anything in the case of nested disjunctions.

emma58 committed:
Merging bigm changes

emma58 committed:
Moving the XOR constraint to be on the transformation block because why not, we don't all have to be like bigm

emma58 committed:
Making chull map disjunctionDatas to the xor constraint, adding error messages to match bigm, adding a million tests, mostly stollen from bigm

emma58 committed:
Oops adding forgotten deletion of an import that doesn't exist

emma58 committed:
All changes to comments...

emma58 committed:
Updating mappings for transformed constraints to also map between constraintDatas as well as the containers... This leads to a lot of complications the largest of which is that single constraints can map to more than one transformed constraint in chull since lb <= expr <= ub structure is allowed.

emma58 committed:
Fixing my stupid error with the apply_to and create_using test by resetting the seed.'

emma58 committed:
Merging from bigm rewrite

emma58 committed:
Testing that variables are always treated as global regardless of where they are declared. However, we don't disaggregate variables which only appear on one disjunct in a disjunction

emma58 committed:
Updating LP file baselines because the XOR constraints and disaggregation constraints have changed name

emma58 committed:
Change to disaggregate fixed variables, cleaning up comments with lies about differences with bigm transformation

emma58 committed:
Removing another comment which is now a lie

emma58 committed:
Beginning to consolidate tests between bigm and chull, catches bug where chull didn't deactivate constraints it transformed

emma58 committed:
More progress on common tests

emma58 committed:
Merging in bigm fix to use same transformation block for transformation of another disjunction from same container...

emma58 committed:
More consolidating of bigm and chull tests

emma58 committed:
Merging master (fixing conflicts with bigm rewrite

michaelbynum committed:
numerical issues in fbbt.interval.asin and fbbt.interval.acos

Robbybp committed:
comments

emma58 committed:
Merge branch 'master' into chull-rewrite

Commit:
:bug: Fix tightening logic when sign is negative

- The logic is being reversed twice; `has_lb()` and `has_ub()` have swapped as well as the setting of `LB` and `UB`

Commit:
:hammer: Use `zip` instead of `enumerate`

Commit:
:sparkles: Add rounding capability

emma58 committed:
Merging master

emma58 committed:
Removing all special handling for local variables in chull. Unless we want to search through the whole model to verify that a locally declared variable on a disjunct is not used anywhere else, we have to disaggregate it in order to avoid silently making mathematical mistakes.

emma58 committed:
Fixing common_tests import

emma58 committed:
Fixing FME test that relies on chull (we now have a transformation block for each disjunction, and that changed the order of the projected constraints again)

emma58 committed:
Removing debugging

emma58 committed:
Merge branch 'master' into chull-rewrite

Robbybp committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo into dae-tools

Robbybp committed:
Functions for identifying and solving for consistent initial conditions, plus tests

emma58 committed:
Moving the methods for finding transformed stuff that are common between bigm and chull to util

emma58 committed:
Moving the warning methods (duplicated code) for active disjuncts and disjunctions to util as well

emma58 committed:
Moving all of bigm's tests that can be common to chull as well to common_tests and adding them to chull tests.

emma58 committed:
Deleting a few unneeded and redundant tests

emma58 committed:
Switching from name to getname in xor method (but leaving name in error messages--I don't care if those are slow)

Robbybp committed:
Addressing reviews

emma58 committed:
Whoops, restoring logic for local variables so that I can at the very least treat inner disjunction disaggregated variables that way.

michaelbynum committed:
Merge branch 'master' into pynumero_interior_point

emma58 committed:
Adding actual test for nested disjunction in chull, fixing the get_transformed_constraints method to always return a list and just complain about a container argument (in both chull and bigm)

emma58 committed:
Adding a LocalVar suffix as the only way to tell chull to not disaggregate a variable.

emma58 committed:
Cleaning up chull, adding tests (some of which are for bigm too)

emma58 committed:
Changing bigm XOR constraint logic to match chull because I don't know why I made it so complicated...

Robbybp committed:
Import and PEP

Robbybp committed:
Merge branch 'dae-tools' of https://github.com/Robbybp/pyomo into dae-tools

Robbybp committed:
Import

emma58 committed:
Testing some error messages for the mapping methods

emma58 committed:
Fixing a bug in bigm where we didn't pick up the BigM suffix if it was on a subblock of a Disjunct, and removing one hack from before the set rewrite merge

emma58 committed:
Adding comments to explain some of what is tested

mrmundt committed:
Trying different PyUtilib branch:

Robbybp committed:
Addressing commments

mrmundt committed:
Merging upstream

emma58 committed:
Fixing BigM Suffixes so that we do look on child blocks of Disjuncts, but only later when we are transforming the constraints

emma58 committed:
Adding name buffer arg to _warn_for_active_disjunct

emma58 committed:
Only passing indexed components through the _transform_disjunct and _transform_disjunction methods, SimpleComponents can go straight to transfrom_somethingData

emma58 committed:
Updates post-set-rewrite, PEP-8ing, switching order of original index and our index in disaggregation constraints (which is why the baselines changed)

emma58 committed:
merging master

michaelbynum committed:
Merge pull request #4 from Robbybp/pynumero_interior_point

solver class, memory reallocation, and regularization

michaelbynum committed:
Merge branch 'master' into pynumero_interior_point

emma58 committed:
Reraising exceptions from the methods to get at the mappings

emma58 committed:
Fixing a typo

jsiirola committed:
Ensure block rules are always called

emma58 committed:
Making how we treat fixed variables in chull an option

jsiirola committed:
Update pyomo.dae now that Blocks always fire their own rules

emma58 committed:
Adding tighter M estimation when user promises that fixed vars will remain so

emma58 committed:
Removing debugging

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo

michaelbynum committed:
interior_point: utilizing a linear solver results object

michaelbynum committed:
a little pynumero reorganization

michaelbynum committed:
addressing issue with mumps interface

michaelbynum committed:
reorganizing pynumero slightly

michaelbynum committed:
Merge branch 'pynumero_linalg' into pynumero_interior_point

michaelbynum committed:
updating tests and imports

michaelbynum committed:
Merge branch 'pynumero_linalg' into pynumero_interior_point

jsiirola committed:
Do not 'densify' indexed blocks with no rule

jsiirola committed:
Update Complementarity component

- track changes in Block's handling of rules
- improve consistency handling Complementarity.Skip in set_value
- incorporate Initializer and disable_methods from core.base.util

michaelbynum committed:
updating imports

michaelbynum committed:
updating tests

michaelbynum committed:
Merge branch 'pynumero_linalg' into pynumero_interior_point

blnicho committed:
Fixing typos

mrmundt committed:
Merge branch 'master' of https://github.com/mrmundt/pyomo

emma58 committed:
Merge branch 'master' into chull-rewrite

michaelbynum committed:
interior point cleanup (mostly uniform linear solver API)

michaelbynum committed:
interior point cleanup

michaelbynum committed:
interior point cleanup

michaelbynum committed:
interior point cleanup

michaelbynum committed:
interior point cleanup

michaelbynum committed:
interior point cleanup

michaelbynum committed:
working on interior point timing

michaelbynum committed:
Interior Point: simplifying evaluations

michaelbynum committed:
simplifying evals

michaelbynum committed:
interior point cleanup

michaelbynum committed:
pynumero linalg cleanup

michaelbynum committed:
Merge branch 'pynumero_linalg' into ip_evals

michaelbynum committed:
interior point cleanup

michaelbynum committed:
Merge pull request #6 from Pyomo/master

merge master into pynumero_interior_point

michaelbynum committed:
Merge remote-tracking branch 'robbybp/pynumero_interior_point' into pynumero_interior_point

Robbybp commented:
Changes look great at a quick scan. I'll try to look through this in detail and play around with it tomorrow evening.

blnicho committed:
Merge pull request #1410 from Robbybp/dae-tools

Solving for consistent initial conditions

blnicho committed:
Fixing typo

michaelbynum committed:
interior point: some performance improvements

DLWoodruff committed:
Updating tests for inverse reduced hessian

michaelbynum committed:
interior point: updating tests

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo

michaelbynum committed:
interior point: adding a method to load variable values back into the pyomo model

mrmundt committed:
Changing badge to GA

mrmundt committed:
Fixing badge path

mrmundt committed:
Trying without the label

michaelbynum committed:
interior point: updating imports

mrmundt committed:
Removing appveyor YML file

michaelbynum committed:
interior point: updating tests

michaelbynum committed:
Merge pull request #1439 from michaelbynum/pynumero_linalg

Pynumero Linalg Updates

michaelbynum committed:
Merge branch 'master' into pynumero_interior_point

mrmundt committed:
Moving around some README stuff

jsiirola committed:
Prevent codecov notification until all builds complete

jsiirola committed:
Include coverage report in Travis build log

michaelbynum committed:
Merge pull request #1438 from jsiirola/block-rule-init

Ensure block rules are always called

michaelbynum committed:
Merge pull request #1360 from michaelbynum/fbbt

FBBT Numerical Issues

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo

mrmundt committed:
Adding FICO Xpress via pip/conda installs to GA Workflows

jsiirola committed:
Merge pull request #1447 from mrmundt/remove_appveyor

Remove Appveyor testing

jsiirola committed:
Merge pull request #1451 from jsiirola/travis-codecov-updates

Coverage configuration updates

jsiirola committed:
Prevent exception for transforamtions missing doc

Transformations that fail to define a doc string should not cause `pyomo
help --transformations` to generate an exception.  This also adds a basic test
that help_transformations() runs and generates reasonable output.

jsiirola committed:
Fix typos in .codecov.yml

The 'notify:' section needs to by under 'codecov:' and not 'coverage:'

jsiirola committed:
Merge pull request #1452 from mrmundt/xpress

Add Xpress to Github Actions

jsiirola committed:
Merge pull request #1454 from jsiirola/fix-help-transformations

Prevent exception for transformations missing doc string

blnicho committed:
Merge pull request #1455 from jsiirola/codecov-config

Fix typos in .codecov.yml

jsiirola committed:
Disable travis use of S3 codecov uploader

jsiirola committed:
Use @wraps so GDP transformation methods inherit docstrings

jsiirola committed:
Quote component names in exception/logger messsages

...also remove trailing whitespace

jsiirola committed:
Make while look compatible with OSX

jsiirola committed:
Make codecov.sh exit with 1 on error

jsiirola committed:
Increase codecov upload retries to 4 with londer delay

jsiirola committed:
Add codecov upload retries to travis builds

jsiirola committed:
Reduce the number of reports before publishing coverage data

jsiirola committed:
Merge pull request #1421 from emma58/chull-rewrite

Rewrite of the GDP Chull transformation

jsiirola committed:
Merge pull request #1457 from jsiirola/update-travis-codecov

Coverage reporting updates

qtothec committed:
revert implementation of rounding, which may lead to improper exclusion of feasible solution points

qtothec committed:
add deprecation warning

jsiirola committed:
Merge pull request #1458 from qtothec/fix_pyomo_constraint_tightener

Fix and deprecate constraint tightener transformation

michaelbynum committed:
Merge branch 'master' into pynumero_interior_point

michaelbynum committed:
interior point: updating imports

michaelbynum committed:
flushing out abstract methods for interior point interface

michaelbynum committed:
adding some pyomo-based method to the interior point interface

Robbybp commented:
> I was having trouble with the the second half of the regularization test. I was getting errors because the regularization coefficient was getting to large, so I cut the second half of that test (the first half was perfect), and added a second test.
 
Before checking out your branch to test/inspect, I pulled from pyomo:master and ran these tests, and got the same error. Not sure what could have changed that is making the algorithm unable to construct or Mumps unable to recognize a matrix with correct inertia.

Robbybp commented:
> Before checking out your branch to test/inspect, I pulled from pyomo:master and ran these tests, and got the same error. 

My scipy regularization test seems to be working fine, FWIW.

michaelbynum commented:
This test problem is really ugly. Without a line search, the algorithm takes steps that are crazy. In the second or third iteration, I see infeasibilities around 1e50. Given the enormous values, there are probably numerical issues that make it very difficult to predict what is going to happen.

Robbybp commented on pyomo/contrib/interior_point/linalg/tests/test_linear_solvers.py:
```diff
@@ -0,0 +1,120 @@
+        self._test_linear_solvers(solver)
+
+
+@unittest.skip('This does not work yet')
+class TestWrongNonzeroOrdering(unittest.TestCase):
+    def _test_solvers(self, solver, use_tril):
+        mat = get_base_matrix(use_tril=use_tril)
+        wrong_order_mat = get_base_matrix_wrong_order(use_tril=use_tril)
+        stat = solver.do_symbolic_factorization(mat)
+        stat = solver.do_numeric_factorization(wrong_order_mat)
+        x_true = np.array([1, 2, 3], dtype=np.double)
+        rhs = mat * x_true
+        x = solver.do_back_solve(rhs)
+        self.assertTrue(np.allclose(x, x_true))
```
What is the purpose of this test?

Robbybp commented:
> This test problem is really ugly. Without a line search, the algorithm takes steps that are crazy. In the second or third iteration, I see infeasibilities around 1e50. Given the enormous values, there are probably numerical issues that make it very difficult to predict what is going to happen.

Yep that matches my observations. Not surprising, I guess, since this is basically the textbook example of degenerate equality constraints.

Robbybp commented:
Would you be opposed if I changed the FactorizationContext to log to a `linear_solver_log_file` as long as one is provided? 

Just read and tested the code and it looks good to me. What do you plan to work on next, out of curiosity?
